indra/llcharacter/llik.cpp

```diff
@@ -3092,55 +3089,92 @@ void LLIK::Solver::executeFabrikPass()
     DEBUG_SET_CONTEXT(outward);
     for (const auto& data_pair : mChainMap)
     {
-        const joint_list_t& chain = data_pair.second;
+        const Joint::joint_vec_t& chain = data_pair.second;
         executeFabrikOutward(chain);
     }
+    // Note: at the end of all this: both local- and world-frame Joint transforms are correct.
+
+    enforceConstraints();
 }
 
-void LLIK::Solver::enforceConstraintsOutward()
+void LLIK::Solver::enforceConstraints()
 {
-    for (const auto& data_pair : mChainMap)
+    DEBUG_SET_CONTEXT(enforce_constraints);
+    // enforce the constraints from outside-->in
+    // this is important because some constraints will also adjust the parent joint
+    // and we need the parent joint's world-frame transform to be correct when that happens
+    std::vector<S16> adjusted_joint_ids;
+    for (chain_map_t::const_reverse_iterator itr = mChainMap.rbegin(); itr != mChainMap.rend(); ++itr)
     {
-        const joint_list_t& chain = data_pair.second;
+        S16 joint_id = enforceConstraintsInward(itr->second);
+        if (joint_id > -1)
+        {
+            // as we proceed we remember the smallest joint_id that was updated on that chain
+            adjusted_joint_ids.push_back(joint_id);
+        }
+    }
```

Comment:
It is inefficient to recompute world-frame transforms that haven't changed, so we spend some CPU cycles above to locate only the joints that have changed (above) and then recompute their world-frame transforms in correct order (below).